asdf_compression/blsc/compressor.py

```diff
@@ -7,10 +7,34 @@ class BloscCompressor(Compressor):
     def compress(self, data, **kwargs):
         import blosc
 
-        yield blosc.compress(data, **kwargs)
+        # Type size, necessary for shuffle filters
+        typesize = data.itemsize
+
+        # Compression level
+        clevel = kwargs.get("level", 9)
+
+        # Shuffle filter
+        shuffle_string = kwargs.get("filter", "bitshuffle")
+        if shuffle_string == "noshuffle":
+            shuffle = blosc.NOSHUFFLE
+        elif shuffle_string == "byteshuffle":
+            shuffle = blosc.BYTESHUFFLE
+        else:
+            # Fallback and default
+            shuffle = blosc.BITSHUFFLE
+
+        # CODEC name
+        cname = kwargs.get("codec", "blosclz")
+
+        # Number of threads
+        nthreads = kwargs.get("threads", 1)
+        self._api.set_nthreads(nthreads)
+
+        yield blosc.compress(data, typesize=typesize, clevel=clevel, shuffle=shuffle, cname=cname, **kwargs)
 
     def decompress(self, data, out, **kwargs):
         import blosc
 
+        # TODO: call `self._api.decompress_ptr` instead to avoid copying the output
```

Comment:
Great idea! I wasn't sure how to do this with the current structure (where data is an iterator that spits out chunks). Is there a way to iteratively decompress with blosc or does it need access to the full data to decompress (I'm not at all familiar with how it works)?

Reply:
Internally `blosc` supports all the things you mention. The Python API is very limited and only supports two decompression calls – one that allocates (and we copy the result) and one that stores via a given pointer.

Since `asdf` seems to want to allocate the output buffer ahead of time, the second API would be nice to use. For this we need to convert `asdf`'s pre-allocated buffer into a pointer. I don't know enough Python for this.

Alternatively we could change `asdf`'s `decompress` API to not take the `out` buffer as input, but rather as the `decompress` function to allocate the output buffer.

How `blosc`'s decompression routine processes its input is unrelated. The Python API doesn't support reading in chunks.

`blosc` compression and decompression are limited in that they can handle at most 2 GByte. `blosc2` handles arbitrary sizes, but there isn't any Python package for it available yet. That Python package should probably support chunked compression and decompression, the `blosc2` C library supports it.

Reply:
Thanks! I like your idea of allowing decompress to allocate the buffer. This combined with passing perhaps the file pointer (instead of chunks read out of the file) might allow blosc to use `decompress_ptr` and avoid the copy of both the input chunks and the output buffer. I think these would be only minor changes in asdf but I'd have to look in more detail to be sure.